src/node_api.cc

```diff
@@ -0,0 +1,2430 @@
+                              void* data,
+                              int property_count,
+                              const napi_property_descriptor* properties,
+                              napi_value* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  v8::Isolate* isolate = v8impl::V8IsolateFromJsEnv(e);
+
+  v8::EscapableHandleScope scope(isolate);
+  v8::Local<v8::Object> cbdata =
+      v8impl::CreateFunctionCallbackData(e, constructor, data);
+
+  RETURN_STATUS_IF_FALSE(!cbdata.IsEmpty(), napi_generic_failure);
+
+  v8::Local<v8::FunctionTemplate> tpl = v8::FunctionTemplate::New(
+      isolate, v8impl::FunctionCallbackWrapper::Invoke, cbdata);
+
+  // we need an internal field to stash the wrapped object
+  tpl->InstanceTemplate()->SetInternalFieldCount(1);
+
+  v8::Local<v8::String> namestring;
+  CHECK_NEW_FROM_UTF8(isolate, namestring, utf8name);
+  tpl->SetClassName(namestring);
+
+  int staticPropertyCount = 0;
+  for (int i = 0; i < property_count; i++) {
+    const napi_property_descriptor* p = properties + i;
+
+    if ((p->attributes & napi_static_property) != 0) {
+      // Static properties are handled separately below.
+      staticPropertyCount++;
+      continue;
+    }
+
+    v8::Local<v8::String> propertyname;
+    CHECK_NEW_FROM_UTF8(isolate, propertyname, p->utf8name);
+
+    v8::PropertyAttribute attributes =
+        static_cast<v8::PropertyAttribute>(p->attributes);
+
+    // This code is similar to that in napi_define_property(); the
+    // difference is it applies to a template instead of an object.
+    if (p->method) {
+      v8::Local<v8::Object> cbdata =
+          v8impl::CreateFunctionCallbackData(e, p->method, p->data);
+
+      RETURN_STATUS_IF_FALSE(!cbdata.IsEmpty(), napi_generic_failure);
+
+      v8::Local<v8::FunctionTemplate> t =
+          v8::FunctionTemplate::New(isolate,
+                                    v8impl::FunctionCallbackWrapper::Invoke,
+                                    cbdata,
+                                    v8::Signature::New(isolate, tpl));
+      t->SetClassName(propertyname);
+
+      tpl->PrototypeTemplate()->Set(propertyname, t, attributes);
+    } else if (p->getter || p->setter) {
+      v8::Local<v8::Object> cbdata =
+          v8impl::CreateAccessorCallbackData(e, p->getter, p->setter, p->data);
+
+      tpl->PrototypeTemplate()->SetAccessor(
+          propertyname,
+          v8impl::GetterCallbackWrapper::Invoke,
+          p->setter ? v8impl::SetterCallbackWrapper::Invoke : nullptr,
+          cbdata,
+          v8::AccessControl::DEFAULT,
+          attributes);
+    } else {
+      v8::Local<v8::Value> value = v8impl::V8LocalValueFromJsValue(p->value);
+      tpl->PrototypeTemplate()->Set(propertyname, value, attributes);
+    }
+  }
+
+  *result = v8impl::JsValueFromV8LocalValue(scope.Escape(tpl->GetFunction()));
+
+  if (staticPropertyCount > 0) {
+    std::vector<napi_property_descriptor> staticDescriptors;
```

Comment:
That's fair. We haven't been very strict about consistency here. Couple questions before I dig in and rename all of these. :)

- I assume snake_case should also be used for all parameters, function names, etc? IIRC, some parameter names are camelCased right now.
- Are there rules around compound words? Is ```int bufsize``` allowed, for example, or should it always be ```int buf_size``` or ```int buf_length```?
- Leading underscores in names like ```_this```? 
- Are there casing rules for classes and class methods? For example, should we rename ```CallbackWrapper``` or ```CallbackWrapper::SetReturnValue``` or ```CallbackWrapper::_argsLength```?